zerver/lib/message.py

```diff
@@ -1697,3 +1697,109 @@ def update_to_dict_cache(
 
     cache_set_many(items_for_remote_cache)
     return message_ids
+
+
+def visibility_policy_for_participation(
+    sender: UserProfile,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a user
+    participates in a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    if sender.is_bot:
+        return None
+
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.UNMUTED
+
+    return None
+
+
+def visibility_policy_for_send_message(
+    sender: UserProfile,
+    message: Message,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a message
+    is sent to a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    # We prioritize 'FOLLOW' over 'UNMUTE'.
+    # We need to carefully handle the following two cases:
+    #
+    # 1. When a user starts a topic
+    #    note: starting a topic is INITIATION as well as PARTICIPATION
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |           ON_INITIATION              |     FOLLOWED
+    #         ON_INITIATION                |          ON_PARTICIPATION            |     FOLLOWED
+    #         ON_INITIATION                |            ON_INITIATION             |     FOLLOWED
+    #
+    # 2. When both the policies have the same value 'ON_PARTICIPATION'.
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |          ON_PARTICIPATION            |     FOLLOWED
+    visibility_policy = None
+
+    if sender.is_bot:
+        return visibility_policy
+
+    visibility_policy = visibility_policy_for_participation(sender)
+
+    if visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+        return visibility_policy
+
+    accessible_messages_in_topic = UserMessage.objects.filter(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    )
```

Comment:
This query is probably more expensive than would be ideal -- if there's 800 messsages in the topic, we end up asking the database to count them all.

In theory all we need is `.exclude(id=message.id).exists()` for the below calculations.

Reply:
Yes, using `.exclude(id=message.id).exists()` is far better. Done.